proving/worker.go

```diff
@@ -0,0 +1,193 @@
+package proving
+
+import (
+	"bytes"
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"fmt"
+	"io"
+	"sort"
+	"sync"
+	"unsafe"
+
+	"github.com/spacemeshos/post/oracle"
+	"github.com/spacemeshos/post/shared"
+)
+
+type batch struct {
+	Data    []byte
+	Index   uint64
+	Release func()
+}
+
+type solution struct {
+	Nonce uint
+	Index uint64
+}
+
+// ioWorker is a worker that reads labels from disk and writes them to a batch channel to be processed by the
+// labelWorkers.
+//
+// TODO(mafa): use this as base to replace GranSpecificReader / GranSpecificWriter and the persistence package.
+func ioWorker(ctx context.Context, batchChan chan<- *batch, b uint32, source io.ReadCloser) error {
+	defer close(batchChan)
+	defer source.Close()
+	index := uint64(0)
+
+	batchDataPool := sync.Pool{
+		New: func() any {
+			buf := make([]byte, BlocksPerWorker*b)
+			return &buf
+		},
+	}
+
+	for {
+		data := *(batchDataPool.Get().(*[]byte))
+		n, err := source.Read(data)
+		switch {
+		case err == io.EOF || err == nil:
+			// make sure we don't send partial blocks to the label workers.
+			// TODO(mafa): this silently drops partial blocks and should be handled better.
+			n -= n % int(b)
+			batch := &batch{
+				Data:  data[:n],
+				Index: index,
+				Release: func() {
+					batchDataPool.Put(&data)
+				},
+			}
+			select {
+			case batchChan <- batch:
+				if err == io.EOF {
+					return nil
+				}
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		default:
+			data = data[:n]
+			batchDataPool.Put(&data)
```

Comment:
Why reslice `data` here?

```suggestion
			batchDataPool.Put(&data)
```